src/test/java/uk/gov/hmcts/reform/iacasedocumentsapi/domain/handlers/presubmit/bail/BailEndApplicationCreatorTest.java

```diff
@@ -0,0 +1,109 @@
+package uk.gov.hmcts.reform.iacasedocumentsapi.domain.handlers.presubmit.bail;
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import uk.gov.hmcts.reform.iacasedocumentsapi.domain.entities.BailCase;
+import uk.gov.hmcts.reform.iacasedocumentsapi.domain.entities.BailCaseFieldDefinition;
+import uk.gov.hmcts.reform.iacasedocumentsapi.domain.entities.DocumentTag;
+import uk.gov.hmcts.reform.iacasedocumentsapi.domain.entities.ccd.CaseDetails;
+import uk.gov.hmcts.reform.iacasedocumentsapi.domain.entities.ccd.Event;
+import uk.gov.hmcts.reform.iacasedocumentsapi.domain.entities.ccd.callback.Callback;
+import uk.gov.hmcts.reform.iacasedocumentsapi.domain.entities.ccd.callback.PreSubmitCallbackResponse;
+import uk.gov.hmcts.reform.iacasedocumentsapi.domain.entities.ccd.callback.PreSubmitCallbackStage;
+import uk.gov.hmcts.reform.iacasedocumentsapi.domain.entities.ccd.field.Document;
+import uk.gov.hmcts.reform.iacasedocumentsapi.domain.service.BailDocumentHandler;
+import uk.gov.hmcts.reform.iacasedocumentsapi.domain.service.DocumentCreator;
+
+@ExtendWith(MockitoExtension.class)
+class BailEndApplicationCreatorTest {
+
+    @Mock private DocumentCreator<BailCase> bailDocumentCreator;
+    @Mock private BailDocumentHandler bailDocumentHandler;
+    @Mock private Callback<BailCase> callback;
+    @Mock private CaseDetails<BailCase> caseDetails;
+    @Mock private BailCase bailCase;
+    @Mock private Document bailEndApplicationDocument;
+
+    private BailEndApplicationCreator bailEndApplicationCreator;
+
+    @BeforeEach
+    void setUp() {
+        bailEndApplicationCreator =
+            new BailEndApplicationCreator(bailDocumentCreator, bailDocumentHandler);
+    }
+
+    @Test
+    void should_handle_valid_events_stage() {
+        for (Event event : Event.values()) {
+            when(callback.getEvent()).thenReturn(event);
+            for (PreSubmitCallbackStage preSubmitCallbackStage : PreSubmitCallbackStage.values()) {
+                when(callback.getCaseDetails()).thenReturn(caseDetails);
+                when(caseDetails.getCaseData()).thenReturn(bailCase);
```

Comment:
When you removed the caseDetails and bailCase mentioned in the above comment, you will need to remove these too.